subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/SkipEmptySourceFilesTaskExecuter.java

```diff
@@ -62,6 +53,7 @@ public TaskExecuterResult execute(TaskInternal task, TaskStateInternal state, fi
         TaskProperties taskProperties = context.getTaskProperties();
         FileCollection sourceFiles = taskProperties.getSourceFiles();
         if (taskProperties.hasSourceFiles() && sourceFiles.isEmpty()) {
+            taskInputsListener.onExecute(task, Cast.cast(FileCollectionInternal.class, sourceFiles));
             AfterPreviousExecutionState previousExecution = context.getAfterPreviousExecution();
             @SuppressWarnings("RedundantTypeArguments")
             ImmutableSortedMap<String, FileCollectionFingerprint> outputFiles = previousExecution == null
```

Comment:
I think we should run the task if the previous execution is null - so we have one execution with empty sources and no outputs. So the state in the execution history is the same if the sources were always empty of if they became empty.